### ISSUE TITLE
 Is there a l2 loss layer in mxnet ?

### ISSUE THREAD

**Issue by brightown** [apache/incubator-mxnet#784]
Does mxnet support l2 loss now?Like EuclideanLossLayer in caffe?
If I want to add other loss function, e.g. hinge loss,how to implement it?

**Comment by jermainewang** [comment]
Hi, there are several ways to do that:
* If you wanna quick prototyping, use NumpyOp to implement it. The example is [here](https://github.com/dmlc/mxnet/blob/master/example/numpy-ops/numpy_softmax.py).
* If you are a python guy and do not want to touch C++ side, you could use MXNet RTC. The example is [here](https://github.com/dmlc/mxnet/blob/master/example/numpy-ops/ndarray_softmax.py). (Well, you still need to know how to write GPU codes).
* If you have your own c++ implementation. Follow the instruction [here](http://mxnet.readthedocs.org/en/latest/tutorial/new_op_howto.html) on how to implement new operators in mxnet.

And we warmly welcome you merge it back when you have implemented it.

-Minjie

**Comment by antinucleon** [comment]
There is regression output op, which use L2 as loss

**Comment by tqchen** [comment]
c.f. http://mxnet.readthedocs.org/en/latest/python/symbol.html#mxnet.symbol.LinearRegressionOutput
This page contains all the symbolic operators, maybe we should put the reference to ops to a more clear place

**Comment by suhangpro** [comment]
Seems the usage of `LinearRegressionOutput` and `LogisticRegressionOutput` is not clearly documented. In my understanding, '...Output' are designed to be the output layers of the network, and behave differently at training and predicting. E.g. for `SoftmaxOutput`:
* At training: first softmax, and then a cross-entropy loss function attached automatically
* At predicting: softmax only, outputs values in [0,1]

We see here `SoftmaxOut` is doing softmax as the name suggested, but `LinearRegressionOut` seems do nothing other than the attached L2 loss at training time: 
```python
data = mx.symbol.Variable('data')
label = mx.symbol.Variable('gt_label')
lr = mx.symbol.LinearRegressionOutput(data=data,label=label,name='lr')
c_exec=lr.bind(ctx=mx.cpu(),args={'data':mx.nd.ones((2,3)),'gt_label':mx.nd.ones((2,3))})
c_exec.forward()
c_exec.outputs[0].asnumpy() # --> array([[ 1., 1., 1.], [ 1., 1., 1.]], dtype=float32)
lr.list_arguments()         # --> ['data', 'gt_label']
```
Some would expect from the name that `LinearRegressionOutput` will insert a linear regression operation (essentially a fully-connected layer). 

Also, `LogisticRegressionOutput` is only doing a logistic function 1/(1+exp(-x)) at forward: 
```python
data = mx.symbol.Variable('data')
label = mx.symbol.Variable('gt_label')
lr = mx.symbol.LogisticRegressionOutput(data=data,label=label,name='lr')
c_exec=lr.bind(ctx=mx.cpu(),args={'data':mx.nd.ones((2,3)),'gt_label':mx.nd.ones((2,3))})
c_exec.forward()
c_exec.outputs[0].asnumpy() # --> array([[ 0.7310586, 0.7310586, 0.7310586], [ 0.7310586, 0.7310586, 0.7310586]], dtype=float32)
lr.list_arguments()         # --> ['data', 'gt_label']
```
Am I getting anything wrong? Also, what's the loss function used for `LogisticRegressionOutput`? Cross-entropy loss still?

**Comment by brightown** [comment]
thank you for all your reply.
I saw c++ code in src\operator,regression_output-inl.h,regression_output.cc and their reference file,but I cannot understand why to calculate the euclidean loss,is it happened in mshadow?

**Comment by brightown** [comment]
thank you for all your reply.
I saw c++ code in src\operator,regression_output-inl.h,regression_output.cc and their reference file,but I cannot understand why to calculate the euclidean loss,is it happened in mshadow?

**Comment by brightown** [comment]
thank you for all your reply.
I saw c++ code in src\operator,regression_output-inl.h,regression_output.cc and their reference file,but I cannot understand why to calculate the euclidean loss,is it happened in mshadow?

**Comment by piiswrong** [comment]
@brightown check src/operator/mshadow_op.h

**Comment by EasonD3** [comment]
@suhangpro Not sure if you already got the answers or not. But to your question, according to my understanding, the reason the `*Output` layers are designed in their way is to avoid switching between training and prediction phases (in comparison, Caffe requires one to explicitly set the `phase` to be either TRAIN or TEST). I don't know why the `*Output`layers in MXNet are designed in this way. Is it because it's not trivial to explicitly set the phase in those layers? @tqchen @antinucleon @piiswrong

**Comment by eric-haibin-lin** [comment]
We should update the API doc of `LogisticRegressionOutput ` with what loss is actually used.
